cmd/soroban-rpc/internal/daemon/daemon.go

```diff
@@ -299,28 +300,44 @@ func (d *Daemon) mustInitializeStorage(cfg *config.Config) *feewindow.FeeWindows
 	defer cancelReadTxMeta()
 	var initialSeq uint32
 	var currentSeq uint32
-	dataMigrations, err := db.BuildMigrations(readTxMetaCtx, d.logger, d.db, cfg)
+	applicableRange, err := db.GetMigrationLedgerRange(readTxMetaCtx, d.db, cfg.HistoryRetentionWindow)
 	if err != nil {
-		d.logger.WithError(err).Fatal("could not build migrations")
+		d.logger.WithError(err).Fatal("could not get ledger range for migration")
 	}
 
-	// Merge migrations range and fee stats range to get the applicable range
-	latestLedger, err := db.NewLedgerEntryReader(d.db).GetLatestLedgerSequence(readTxMetaCtx)
+	maxFeeRetentionWindow := max(cfg.ClassicFeeStatsLedgerRetentionWindow, cfg.SorobanFeeStatsLedgerRetentionWindow)
+	feeStatsRange, err := db.GetMigrationLedgerRange(readTxMetaCtx, d.db, maxFeeRetentionWindow)
 	if err != nil {
-		d.logger.WithError(err).Fatal("failed to get latest ledger sequence: %w", err)
+		d.logger.WithError(err).Fatal("could not get ledger range for fee stats")
 	}
 
-	maxFeeRetentionWindow := max(cfg.ClassicFeeStatsLedgerRetentionWindow, cfg.SorobanFeeStatsLedgerRetentionWindow)
-	ledgerSeqRange := &db.LedgerSeqRange{FirstLedgerSeq: latestLedger - maxFeeRetentionWindow, LastLedgerSeq: latestLedger}
-	applicableRange := dataMigrations.ApplicableRange()
-	ledgerSeqRange = ledgerSeqRange.Merge(applicableRange)
+	// Combine the ledger range for fees, events and transactions
+	ledgerSeqRange := feeStatsRange.Merge(applicableRange)
+
+	// Start a common db transaction for the entire migration duration
+	err = d.db.Begin(readTxMetaCtx)
+	if err != nil {
+		d.logger.WithError(err).Fatal("could not commit database transaction: ", d.db.Rollback())
```

Comment:
d.db.Rollback() is passed as an argument to the log message but the log message expects no place holders (e.g. `%v`)

Reply:
@tamirms `Fatal` function does not have any placeholders. We can just chain strings and variables like I have done.

Reply:
I think this approach is a bit unintuitive (why should the rollback error be printed in the log message without any context?) and error prone (if we forget to call rollback as an argument to the fatal call we can accidentally commit the transaction). Consider this alternative:

after creating the transaction, refactor the remaining code in this function to be a separate function which returns an error. The separate function would no longer call `log...Fatal()` upon encountering an error. Instead it would return the error and the code which handles the return value would rollback (just in one place) and call `log...Fatal()` to terminate.